### PR TITLE
Feature/add profile functionality

### PR DESCRIPTION
**Pull request by NathanDVT**
* Added Profile functionality (ViewModel, Repo, ViewController protocol and Model)
* Added Mock ViewModel Instance

### CODE REVIEW

File: NLibrary/Classes/ProfileMVVM/ProfileRepo.swift [hunk 1]
```diff
@@ -0,0 +1,39 @@
+
```
Comment (NathanDVT) [inline NLibrary/Classes/ProfileMVVM/ProfileRepo.swift:1]: Codacy Issue found: [File shouldn't start with whitespace: currently starts with 1 whitespace characters](https://app.codacy.com/manual/NathanDVT/NLibrary/pullRequest?prid=5382838)

**Comment by NathanDVT** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 18
           

Clones added
============
- ProfileMVVMq/ProfileRepo.swift  1
- NLibrary/Classes/ProfileMVVM/ProfileRepo.swift  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/NathanDVT/NLibrary/pullRequest?prid=5382838&bid=17649483)

File: NLibrary/Classes/ProfileMVVM/ProfileRepo.swift [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        do {
+            // TO DO: cancel observers hanging
+            try Auth.auth().signOut()
+            completion(.succesfullRequest)
+        } catch (let error){
+            completion(.failedRequest(message: error.localizedDescription))
+        }
+    }
+    
+
```
Comment (NathanDVT) [inline NLibrary/Classes/ProfileMVVM/ProfileRepo.swift:38]: Codacy Issue found: [Lines should not have trailing whitespace.](https://app.codacy.com/manual/NathanDVT/NLibrary/pullRequest?prid=5382838)

File: ProfileMVVMq/ProfileViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,20 @@
+
```
Comment (NathanDVT) [inline ProfileMVVMq/ProfileViewModel.swift:1]: Codacy Issue found: [File shouldn't start with whitespace: currently starts with 1 whitespace characters](https://app.codacy.com/manual/NathanDVT/NLibrary/pullRequest?prid=5382838)

File: NLibrary/Classes/ProfileMVVM/Mocks/ProfileMockViewModel.swift [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    
+    public func getUserProfile() {
+        getUserProfileCalled = true
+    }
+    
+    public func logoutRequest() {
+        logoutRequestCalled = true
+    }
+    
+
```
Comment (NathanDVT) [inline NLibrary/Classes/ProfileMVVM/Mocks/ProfileMockViewModel.swift:26]: Codacy Issue found: [Lines should not have trailing whitespace.](https://app.codacy.com/manual/NathanDVT/NLibrary/pullRequest?prid=5382838)

File: NLibrary/Classes/ProfileMVVM/Prototypes/ProfileRepoProtocol.swift [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+//  taylorswift
+//
+//  Created by Nathan Ngobale on 2020/03/20.
+//  Copyright © 2020 Nathan Ngobale. All rights reserved.
+//
+
+import Foundation
+import AVKit
+
+
```
Comment (NathanDVT) [inline NLibrary/Classes/ProfileMVVM/Prototypes/ProfileRepoProtocol.swift:12]: Codacy Issue found: [Limit vertical whitespace to a single empty line. Currently 2.](https://app.codacy.com/manual/NathanDVT/NLibrary/pullRequest?prid=5382838)

File: NLibrary/Classes/ProfileMVVM/ProfileRepo.swift [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+            completion(.success(ProfileModel(dictionary: result)))
+        })
+    }
+
+    public func logoutRequest(completion: @escaping(APIRequestResult) -> Void) {
+        do {
+            // TO DO: cancel observers hanging
+            try Auth.auth().signOut()
+            completion(.succesfullRequest)
+        } catch (let error){
```
Comment (NathanDVT) [inline NLibrary/Classes/ProfileMVVM/ProfileRepo.swift:33]: Codacy Issue found: [`if`, `for`, `guard`, `switch`, `while`, and `catch` statements shouldn't unnecessarily wrap their conditionals or arguments in parentheses.](https://app.codacy.com/manual/NathanDVT/NLibrary/pullRequest?prid=5382838)

File: ProfileMVVMq/ProfileModel.swift [hunk 7]
```diff
@@ -0,0 +1,42 @@
+
```
Comment (NathanDVT) [inline ProfileMVVMq/ProfileModel.swift:1]: Codacy Issue found: [File shouldn't start with whitespace: currently starts with 1 whitespace characters](https://app.codacy.com/manual/NathanDVT/NLibrary/pullRequest?prid=5382838)

File: NLibrary/Classes/ProfileMVVM/ProfileViewModel.swift [hunk 8]
```diff
@@ -0,0 +1,32 @@
+
```
Comment (NathanDVT) [inline NLibrary/Classes/ProfileMVVM/ProfileViewModel.swift:1]: Codacy Issue found: [File shouldn't start with whitespace: currently starts with 1 whitespace characters](https://app.codacy.com/manual/NathanDVT/NLibrary/pullRequest?prid=5382838)

File: NLibrary/Classes/Service Files/Mocks/FirebaseServiceMock.swift [hunk 9]
```diff
@@ -9,6 +9,10 @@
 import Foundation
 
 public class FirebaseServiceMock: NSObject, FirebaseServiceProtocol {
+    public required override init() {
+        
+    }
+
```
Comment (NathanDVT) [inline NLibrary/Classes/Service Files/Mocks/FirebaseServiceMock.swift:15]: Codacy Issue found: [Lines should not have trailing whitespace.](https://app.codacy.com/manual/NathanDVT/NLibrary/pullRequest?prid=5382838)

File: ProfileMVVMq/ProfileModel.swift [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+        let titName = dictionary.value(forKey: "titleName") as? String,
+        let preUrl = dictionary.value(forKey: "previewUrl") as? String else {
+            return
+        }
+
+        gender = .male
+        self.email = email
+        noPlaylists = 0
+    }
+
```
Comment (NathanDVT) [inline ProfileMVVMq/ProfileModel.swift:41]: Codacy Issue found: [Lines should not have trailing whitespace.](https://app.codacy.com/manual/NathanDVT/NLibrary/pullRequest?prid=5382838)

File: NLibrary/Classes/ProfileMVVM/Prototypes/ProfileViewControllerProtocol.swift [hunk 11]
```diff
@@ -0,0 +1,8 @@
+
```
Comment (NathanDVT) [inline NLibrary/Classes/ProfileMVVM/Prototypes/ProfileViewControllerProtocol.swift:1]: Codacy Issue found: [File shouldn't start with whitespace: currently starts with 1 whitespace characters](https://app.codacy.com/manual/NathanDVT/NLibrary/pullRequest?prid=5382838)

File: ProfileMVVMq/ProfileRepo.swift [hunk 12]
```diff
@@ -0,0 +1,45 @@
+
```
Comment (NathanDVT) [inline ProfileMVVMq/ProfileRepo.swift:1]: Codacy Issue found: [File shouldn't start with whitespace: currently starts with 1 whitespace characters](https://app.codacy.com/manual/NathanDVT/NLibrary/pullRequest?prid=5382838)

File: NLibrary/Classes/DashboardHomeMVVM/DashboardViewModel.swift [hunk 13]
```diff
@@ -2,9 +2,9 @@
 import Foundation
 import AVKit
 
+
```
Comment (NathanDVT) [inline NLibrary/Classes/DashboardHomeMVVM/DashboardViewModel.swift:5]: Codacy Issue found: [Limit vertical whitespace to a single empty line. Currently 2.](https://app.codacy.com/manual/NathanDVT/NLibrary/pullRequest?prid=5382838)

File: NLibrary/Classes/ProfileMVVM/ProfileRepo.swift (hunk 6 above)
Comment (NathanDVT) [inline NLibrary/Classes/ProfileMVVM/ProfileRepo.swift:33]: Codacy Issue found: [Opening braces should be preceded by a single space and on the same line as the declaration.](https://app.codacy.com/manual/NathanDVT/NLibrary/pullRequest?prid=5382838)

File: NLibrary/Classes/ProfileMVVM/ProfileRepo.swift (hunk 2 above)
Comment (NathanDVT) [inline NLibrary/Classes/ProfileMVVM/ProfileRepo.swift:38]: Codacy Issue found: [Limit vertical whitespace to a single empty line. Currently 2.](https://app.codacy.com/manual/NathanDVT/NLibrary/pullRequest?prid=5382838)

File: NLibrary/Classes/ProfileMVVM/ProfileModel.swift [hunk 14]
```diff
@@ -0,0 +1,54 @@
+
```
Comment (NathanDVT) [inline NLibrary/Classes/ProfileMVVM/ProfileModel.swift:1]: Codacy Issue found: [File shouldn't start with whitespace: currently starts with 1 whitespace characters](https://app.codacy.com/manual/NathanDVT/NLibrary/pullRequest?prid=5382838)

File: NLibrary/Classes/Service Files/FirebaseService.swift [hunk 15]
```diff
@@ -23,6 +27,9 @@ public enum APIRequestResult: Error {
     private var repoDashBoard: DashboardRepoProtocol?
     private var repoPlaylist: PlaylistRepo?
 
+    public required override init() {
+
```
Comment (NathanDVT) [inline NLibrary/Classes/Service Files/FirebaseService.swift:31]: Codacy Issue found: [Lines should not have trailing whitespace.](https://app.codacy.com/manual/NathanDVT/NLibrary/pullRequest?prid=5382838)

File: NLibrary/Classes/ProfileMVVM/Mocks/ProfileMockViewModel.swift (hunk 4 above)
Comment (NathanDVT) [inline NLibrary/Classes/ProfileMVVM/Mocks/ProfileMockViewModel.swift:26]: Codacy Issue found: [Limit vertical whitespace to a single empty line. Currently 2.](https://app.codacy.com/manual/NathanDVT/NLibrary/pullRequest?prid=5382838)

**Comment by NathanDVT** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/NathanDVT/NLibrary/pullRequest?prid=5382838&bid=17649483)

File: NLibrary/Classes/ProfileMVVM/ProfileRepo.swift [hunk 16]
```diff
@@ -0,0 +1,37 @@
+
```
Comment (NathanDVT) [inline NLibrary/Classes/ProfileMVVM/ProfileRepo.swift:1]: Codacy Issue found: [File shouldn't start with whitespace: currently starts with 1 whitespace characters](https://app.codacy.com/manual/NathanDVT/NLibrary/pullRequest?prid=5382838)

**Review by JancoDVTDev — APPROVED** [review]
Looks good, I see commented code, just make sure that if it isn't code you come back to that it be removed

**Pull request merged**